<a href="https://colab.research.google.com/github/athenaap/data_science_hw/blob/main/Homework6V5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Case Intro
Term deposits are a major source of income for a bank. A term deposit is a cash investment held at a financial institution. Your money is invested for an agreed rate of interest over a fixed amount of time, or term. The bank has various outreach plans to sell term deposits to their customers such as email marketing, advertisements, telephonic marketing, and digital marketing.

Telephonic marketing campaigns still remain one of the most effective way to reach out to people. However, they require huge investment as large call centers are hired to actually execute these campaigns. Hence, it is crucial to identify the customers most likely to convert beforehand so that they can be specifically targeted via call.

The data is related to direct marketing campaigns (phone calls) of a Portuguese banking institution. The classification goal is to predict if the client will subscribe to a term deposit (variable y).

Content
The data is related to the direct marketing campaigns of a Portuguese banking institution. The marketing campaigns were based on phone calls. Often, more than one contact to the same client was required, in order to access if the product (bank term deposit) would be ('yes') or not ('no') subscribed by the customer or not. The data folder contains two datasets:-

Bank.csv: 45,211 rows and 18 columns ordered by date (from May 2008 to November 2010)

Detailed Column Descriptions
bank client data:

1 - age (numeric)

2 - job : type of job (categorical: "admin.","unknown","unemployed","management","housemaid","entrepreneur","student",
"blue-collar","self-employed","retired","technician","services")

3 - marital : marital status (categorical: "married","divorced","single"; note: "divorced" means divorced or widowed)

4 - education (categorical: "unknown","secondary","primary","tertiary")

5 - default: has credit in default? (binary: "yes","no")

6 - balance: average yearly balance, in euros (numeric)

7 - housing: has housing loan? (binary: "yes","no")

8 - loan: has personal loan? (binary: "yes","no")
# related with the last contact of the current campaign:
9 - contact: contact communication type (categorical: "unknown","telephone","cellular")
10 - day: last contact day of the month (numeric)

11 - month: last contact month of year (categorical: "jan", "feb", "mar", …, "nov", "dec")

12 - duration: last contact duration, in seconds (numeric)

# other attributes:
13 - campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)

14 - pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric, -1 means client was not previously contacted)

15 - previous: number of contacts performed before this campaign and for this client (numeric)

16 - poutcome: outcome of the previous marketing campaign (categorical: "unknown","other","failure","success")

Output variable (desired target):

17 - y - has the client subscribed a term deposit? (binary: "yes","no")

Missing Attribute Values: None


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
df=pd.read_csv('https://raw.githubusercontent.com/ogut77/DataScience/main/data/Bank.csv',sep = ';')
df

age           job   marital  education default  balance housing loan  \
0       58    management   married   tertiary      no     2143     yes   no   
1       44    technician    single  secondary      no       29     yes   no   
2       33  entrepreneur   married  secondary      no        2     yes  yes   
3       47   blue-collar   married    unknown      no     1506     yes   no   
4       33       unknown    single    unknown      no        1      no   no   
...    ...           ...       ...        ...     ...      ...     ...  ...   
45206   51    technician   married   tertiary      no      825      no   no   
45207   71       retired  divorced    primary      no     1729      no   no   
45208   72       retired   married  secondary      no     5715      no   no   
45209   57   blue-collar   married  secondary      no      668      no   no   
45210   37  entrepreneur   married  secondary      no     2971      no   no   

         contact  day month  duration  campaign  pdays  previous poutcome    y  
0        unknown    5   may       261         1     -1         0  unknown   no  
1        unknown    5   may       151         1     -1         0  unknown   no  
2        unknown    5   may        76         1     -1         0  unknown   no  
3        unknown    5   may        92         1     -1         0  unknown   no  
4        unknown    5   may       198         1     -1         0  unknown   no  
...          ...  ...   ...       ...       ...    ...       ...      ...  ...  
45206   cellular   17   nov       977         3     -1         0  unknown  yes  
45207   cellular   17   nov       456         2     -1         0  unknown  yes  
45208   cellular   17   nov      1127         5    184         3  success  yes  
45209  telephone   17   nov       508         4     -1         0  unknown   no  
45210   cellular   17   nov       361         2    188        11    other   no  

[45211 rows x 17 columns]

In [2]:
print(df.shape)
df.info()
df.isnull().sum()

(45211, 17)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45211 entries, 0 to 45210
Data columns (total 17 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   age        45211 non-null  int64 
 1   job        45211 non-null  object
 2   marital    45211 non-null  object
 3   education  45211 non-null  object
 4   default    45211 non-null  object
 5   balance    45211 non-null  int64 
 6   housing    45211 non-null  object
 7   loan       45211 non-null  object
 8   contact    45211 non-null  object
 9   day        45211 non-null  int64 
 10  month      45211 non-null  object
 11  duration   45211 non-null  int64 
 12  campaign   45211 non-null  int64 
 13  pdays      45211 non-null  int64 
 14  previous   45211 non-null  int64 
 15  poutcome   45211 non-null  object
 16  y          45211 non-null  object
dtypes: int64(7), object(10)
memory usage: 5.9+ MB


age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
y            0
dtype: int64

In [3]:
#For object check the data
for cn in df.columns:
  if(df[cn].dtype==object):
    print(df[cn].value_counts())


job
blue-collar      9732
management       9458
technician       7597
admin.           5171
services         4154
retired          2264
self-employed    1579
entrepreneur     1487
unemployed       1303
housemaid        1240
student           938
unknown           288
Name: count, dtype: int64
marital
married     27214
single      12790
divorced     5207
Name: count, dtype: int64
education
secondary    23202
tertiary     13301
primary       6851
unknown       1857
Name: count, dtype: int64
default
no     44396
yes      815
Name: count, dtype: int64
housing
yes    25130
no     20081
Name: count, dtype: int64
loan
no     37967
yes     7244
Name: count, dtype: int64
contact
cellular     29285
unknown      13020
telephone     2906
Name: count, dtype: int64
month
may    13766
jul     6895
aug     6247
jun     5341
nov     3970
apr     2932
feb     2649
jan     1403
oct      738
sep      579
mar      477
dec      214
Name: count, dtype: int64
poutcome
unknown    36959
failure     4901
other  

In [4]:
# Following function converts non-numeric variables (e.g., 'category', 'object') into numeric using label encoding
# Note:
#Label encoding converts categorical values into integer codes. Each unique category is assigned a distinct number, such as 1, 2, 3, 4, etc.
#Example : In our data, education variable has following values:{ 'primary' 'secondary', 'tertiary', 'unknown'}
#and it is mapped { 'primary': 0, 'secondary': 1, 'tertiary': 2, 'unknown': 3}

#  education  education_encoded
#  secondary                  1
#   tertiary                  2
#    primary                  0
#    unknown                  3
#   tertiary                  2
#    primary                  0
#
#Label encoding is generally suitable for tree-based models (e.g., decision trees, random forests, boosting methods).
#However, it may not be appropriate for models where the objective function relies on distance-based calculations, such as neural networks, support vector machines (SVM), or linear regression.
#For nominal features (categories with no intrinsic ordering, e.g., "red," "blue," "green"), label encoding can mislead the model by implying an ordinal relationship where none exists.
#In such cases, one-hot encoding is usually preferred.
#One-hot encoding creates binary columns (dummy variables) for each category in a categorical feature, avoiding the introduction of unintended ordinality.

def Encoder(df):
          from sklearn import preprocessing
          columnsToEncode = list(df.select_dtypes(include=['category','object']))
          le = preprocessing.LabelEncoder()
          for feature in columnsToEncode:
              try:
                  df[feature] = le.fit_transform(df[feature])
              except:
                  print('Error encoding '+feature)
          return df


In [5]:
df=Encoder(df)
df

age  job  marital  education  default  balance  housing  loan  contact  \
0       58    4        1          2        0     2143        1     0        2   
1       44    9        2          1        0       29        1     0        2   
2       33    2        1          1        0        2        1     1        2   
3       47    1        1          3        0     1506        1     0        2   
4       33   11        2          3        0        1        0     0        2   
...    ...  ...      ...        ...      ...      ...      ...   ...      ...   
45206   51    9        1          2        0      825        0     0        0   
45207   71    5        0          0        0     1729        0     0        0   
45208   72    5        1          1        0     5715        0     0        0   
45209   57    1        1          1        0      668        0     0        1   
45210   37    2        1          1        0     2971        0     0        0   

       day  month  duration  campaign  pdays  previous  poutcome  y  
0        5      8       261         1     -1         0         3  0  
1        5      8       151         1     -1         0         3  0  
2        5      8        76         1     -1         0         3  0  
3        5      8        92         1     -1         0         3  0  
4        5      8       198         1     -1         0         3  0  
...    ...    ...       ...       ...    ...       ...       ... ..  
45206   17      9       977         3     -1         0         3  1  
45207   17      9       456         2     -1         0         3  1  
45208   17      9      1127         5    184         3         2  1  
45209   17      9       508         4     -1         0         3  0  
45210   17      9       361         2    188        11         1  0  

[45211 rows x 17 columns]

In [6]:
y = df['y'] #Output
X = df.drop('y',axis=1)
X

age  job  marital  education  default  balance  housing  loan  contact  \
0       58    4        1          2        0     2143        1     0        2   
1       44    9        2          1        0       29        1     0        2   
2       33    2        1          1        0        2        1     1        2   
3       47    1        1          3        0     1506        1     0        2   
4       33   11        2          3        0        1        0     0        2   
...    ...  ...      ...        ...      ...      ...      ...   ...      ...   
45206   51    9        1          2        0      825        0     0        0   
45207   71    5        0          0        0     1729        0     0        0   
45208   72    5        1          1        0     5715        0     0        0   
45209   57    1        1          1        0      668        0     0        1   
45210   37    2        1          1        0     2971        0     0        0   

       day  month  duration  campaign  pdays  previous  poutcome  
0        5      8       261         1     -1         0         3  
1        5      8       151         1     -1         0         3  
2        5      8        76         1     -1         0         3  
3        5      8        92         1     -1         0         3  
4        5      8       198         1     -1         0         3  
...    ...    ...       ...       ...    ...       ...       ...  
45206   17      9       977         3     -1         0         3  
45207   17      9       456         2     -1         0         3  
45208   17      9      1127         5    184         3         2  
45209   17      9       508         4     -1         0         3  
45210   17      9       361         2    188        11         1  

[45211 rows x 16 columns]

In [7]:
from sklearn.model_selection import  train_test_split


X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=17)




Q1)Using  Random Forest,XGBoost, Light GBM and Gradient Boosting Classifier with default parameters (no parameter specifications except random_state) calculate Accuracy on Test data. Which method gives the best accuracy on test data

Q2) Using optuna hyperparmeter optimization technique and 100 trial

 a)find best methods with  parameters  using Cross validation (CV=3) technique for the range of   parameters below. What are the best parameters for the method with highest cross validation accuracy?
 For random forest


  "max_depth"   : trial.suggest_int("max_depth", 2,  X_train.shape[1]),
  "max_features": trial.suggest_int("max_features", 2, X_train.shape[1])

For XGBoost, Light GBM and Gradient Boosting Classifier

  "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1]),
  "learning_rate": trial.suggest_float("learning_rate", 0.001, 0.3,log=True)

where X_train.shape[1] is number of columnns in the train data.

 b)Evaluate the performance of the  method with highest cross validation accuracy on test data.What is the accuracy value? Are there any improvement of the same method with default parameters?


In [8]:
# Q1)
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier

models=[]
models.append(('Random Forest',RandomForestClassifier(random_state=17)))
models.append(('XGBoost',XGBClassifier(random_state=17)))
models.append(('Light GBM',LGBMClassifier(random_state=17)))
models.append(('Gradient Boosting Classifier',GradientBoostingClassifier(random_state=17)))

for name,model in models:
    model = model
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    print(name,accuracy_score(y_test, y_pred))




Random Forest 0.9025037600636999
XGBoost 0.9036538971954349
[LightGBM] [Info] Number of positive: 3940, number of negative: 29968
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.004688 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 988
[LightGBM] [Info] Number of data points in the train set: 33908, number of used features: 16
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.116197 -> initscore=-2.028949
[LightGBM] [Info] Start training from score -2.028949
Light GBM 0.9073697248518092
Gradient Boosting Classifier 0.9018844554543042


# As we can see from the accuracies, all models show approximately same results, which makes a lot of sense. Because, all models except Random Forest are boosting models, which prove to be very effective as results are iteratively improving upon previous models. At the same time, Random Forest achieves comparable performance to the boosting methods in this case, suggesting that the dataset may not heavily favor one ensemble approach over the other. LightGBM achieved the highest accuracy (0.907), however it is not significant difference to make assumption about it.

In [10]:
!pip install optuna

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 386.6/386.6 kB 8.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 231.9/231.9 kB 8.8 MB/s eta 0:00:00


In [11]:
# Q2)
import optuna
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score


def objective_rf(trial):
    params = {
        "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1]),
        "max_features": trial.suggest_int("max_features", 2, X_train.shape[1]),
        "n_estimators": 100,
        "random_state": 42
    }
    model = RandomForestClassifier(**params)
    score = cross_val_score(model, X_train, y_train, cv=3, scoring='accuracy').mean()
    return score

def objective_xgb(trial):
    params = {
        "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1]),
        "learning_rate": trial.suggest_float("learning_rate", 0.001, 0.3, log=True),
        "n_estimators": 100,
        "random_state": 42
    }
    model = XGBClassifier(**params)
    score = cross_val_score(model, X_train, y_train, cv=3, scoring='accuracy').mean()
    return score

def objective_lgbm(trial):
    params = {
        "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1]),
        "learning_rate": trial.suggest_float("learning_rate", 0.001, 0.3, log=True),
        "n_estimators": 100,
        "random_state": 42
    }
    model = LGBMClassifier(**params)
    score = cross_val_score(model, X_train, y_train, cv=3, scoring='accuracy').mean()
    return score

def objective_gbc(trial):
    params = {
        "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1]),
        "learning_rate": trial.suggest_float("learning_rate", 0.001, 0.3, log=True),
        "n_estimators": 100,
        "random_state": 42
    }
    model = GradientBoostingClassifier(**params)
    score = cross_val_score(model, X_train, y_train, cv=3, scoring='accuracy').mean()
    return score



In [ ]:
def optimize_model(objective, model_name):
    study = optuna.create_study(direction="maximize")
    study.optimize(objective, n_trials=100)
    print(f"Best {model_name} params:", study.best_params)
    print(f"Best CV accuracy:", study.best_value)
    return study.best_params

# Optimization
best_rf = optimize_model(objective_rf, "Random Forest")
best_xgb = optimize_model(objective_xgb, "XGBoost")
best_lgbm = optimize_model(objective_lgbm, "LightGBM")
best_gbc = optimize_model(objective_gbc, "Gradient Boosting")

[I 2025-04-15 19:43:15,087] A new study created in memory with name: no-name-9349f607-05e4-4325-ac9d-0b6e534da086
[I 2025-04-15 19:43:39,000] Trial 0 finished with value: 0.9013507891949702 and parameters: {'max_depth': 5, 'max_features': 14}. Best is trial 0 with value: 0.9013507891949702.
[I 2025-04-15 19:43:44,399] Trial 1 finished with value: 0.8920608886685845 and parameters: {'max_depth': 3, 'max_features': 10}. Best is trial 0 with value: 0.9013507891949702.
[I 2025-04-15 19:43:52,241] Trial 2 finished with value: 0.9041819690421384 and parameters: {'max_depth': 10, 'max_features': 4}. Best is trial 2 with value: 0.9041819690421384.
[I 2025-04-15 19:43:56,013] Trial 3 finished with value: 0.8929751158908409 and parameters: {'max_depth': 8, 'max_features': 2}. Best is trial 2 with value: 0.9041819690421384.
[I 2025-04-15 19:44:02,182] Trial 4 finished with value: 0.9007903511662385 and parameters: {'max_depth': 14, 'max_features': 2}. Best is trial 2 with value: 0.904181969042138

In [ ]:
# Initialize models with best parameters
best_lgbm_model = LGBMClassifier(**best_lgbm, random_state=42)
best_lgbm_model.fit(X_train, y_train)

# Predict on test set
y_pred = best_lgbm_model.predict(X_test)
test_accuracy = accuracy_score(y_test, y_pred)

print(f"Optimized LightGBM Test Accuracy: {test_accuracy:.4f}")

# Compare with default LightGBM
default_lgbm = LGBMClassifier(random_state=42)
default_lgbm.fit(X_train, y_train)
default_accuracy = accuracy_score(y_test, default_lgbm.predict(X_test))

print(f"Default LightGBM Test Accuracy: {default_accuracy:.4f}")
print(f"Improvement: {test_accuracy - default_accuracy:.4f}")

For Q3 and Q4 ,use the following data.

In [ ]:
dr=pd.read_csv('https://raw.githubusercontent.com/ogut77/DataScience/main/data/diamond.csv')
dr

Carat Weight              Cut Color Clarity Polish Symmetry Report  \
0             1.10            Ideal     H     SI1     VG       EX    GIA   
1             0.83            Ideal     H     VS1     ID       ID   AGSL   
2             0.85            Ideal     H     SI1     EX       EX    GIA   
3             0.91            Ideal     E     SI1     VG       VG    GIA   
4             0.83            Ideal     G     SI1     EX       EX    GIA   
...            ...              ...   ...     ...    ...      ...    ...   
5995          1.03            Ideal     D     SI1     EX       EX    GIA   
5996          1.00        Very Good     D     SI1     VG       VG    GIA   
5997          1.02            Ideal     D     SI1     EX       EX    GIA   
5998          1.27  Signature-Ideal     G     VS1     EX       EX    GIA   
5999          2.19            Ideal     E     VS1     EX       EX    GIA   

      Price  
0      5169  
1      3470  
2      3183  
3      4370  
4      3171  
...     ...  
5995   6250  
5996   5328  
5997   6157  
5998  11206  
5999  30507  

[6000 rows x 8 columns]

In [ ]:
def Encoder(df):
          from sklearn import preprocessing
          columnsToEncode = list(df.select_dtypes(include=['category','object']))
          le = preprocessing.LabelEncoder()
          for feature in columnsToEncode:
              try:
                  df[feature] = le.fit_transform(df[feature])
              except:
                  print('Error encoding '+feature)
          return df


In [ ]:
dr=Encoder(dr)
dr

Carat Weight  Cut  Color  Clarity  Polish  Symmetry  Report  Price
0             1.10    2      4        2       3         0       1   5169
1             0.83    2      4        3       2         2       0   3470
2             0.85    2      4        2       0         0       1   3183
3             0.91    2      1        2       3         3       1   4370
4             0.83    2      3        2       0         0       1   3171
...            ...  ...    ...      ...     ...       ...     ...    ...
5995          1.03    2      0        2       0         0       1   6250
5996          1.00    4      0        2       3         3       1   5328
5997          1.02    2      0        2       0         0       1   6157
5998          1.27    3      3        3       0         0       1  11206
5999          2.19    2      1        3       0         0       1  30507

[6000 rows x 8 columns]

In [ ]:
y = dr['Price'] #Output
X = dr.drop('Price',axis=1)
X

Carat Weight  Cut  Color  Clarity  Polish  Symmetry  Report
0             1.10    2      4        2       3         0       1
1             0.83    2      4        3       2         2       0
2             0.85    2      4        2       0         0       1
3             0.91    2      1        2       3         3       1
4             0.83    2      3        2       0         0       1
...            ...  ...    ...      ...     ...       ...     ...
5995          1.03    2      0        2       0         0       1
5996          1.00    4      0        2       3         3       1
5997          1.02    2      0        2       0         0       1
5998          1.27    3      3        3       0         0       1
5999          2.19    2      1        3       0         0       1

[6000 rows x 7 columns]

In [ ]:
from sklearn.model_selection import  train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=17)

Q3)Using Linear Regression,Decison Tree Random Forest,XGBoost, Light GBM and Gradient Boosting Classifier with default parameters (no parameter specifications except random_state) calculate R2 statistics on test data. Which method gives the best accuracy on test data

Q4) Using optuna hyperparmeter optimization technique (100 trial)  with Random Forest,XGBoost, Light GBM and Gradient Boosting Regressor


a)find best methods with  parameters  using Cross validation (CV=3) technique for the range of   parameters below. What are the best parameters for the method with highest cross validation R2?

For random forest


  "max_depth"   : trial.suggest_int("max_depth", 2,  X_train.shape[1]),
  "max_features": trial.suggest_int("max_features", 2, X_train.shape[1])

For XGBoost, Light GBM and Gradient Boosting Classifier

  "max_depth": trial.suggest_int("max_depth", 2, X_train.shape[1]),
  "learning_rate": trial.suggest_float("learning_rate", 0.001, 0.3,log=True)

where X_train.shape[1] is number of columnns in the train data.

 b)Evaluate the performance of the  method with highest cross validation R2 on test data. What is the R2 value? Are there any improvement of the same method with default parameters?
